tests/msc2716_test.go

```diff
@@ -457,6 +458,82 @@ func TestBackfillingHistory(t *testing.T) {
+			eventIDsBefore := createMessagesInRoom(t, alice, roomID, 1)
+			eventIdBefore := eventIDsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			// eventIDsAfter
+			createMessagesInRoom(t, alice, roomID, 2)
+
+			// Insert a backfilled event
+			batchSendRes := batchSendHistoricalMessages(
+				t,
+				as,
+				roomID,
+				eventIdBefore,
+				"",
+				createJoinStateEventsForBackfillRequest([]string{virtualUserID}, timeAfterEventBefore),
+				createMessageEventsForBackfillRequest([]string{virtualUserID}, timeAfterEventBefore, 3),
+				// Status
+				200,
+			)
+			batchSendResBody0 := client.ParseJSON(t, batchSendRes)
+			historicalEventIDs0 := client.GetJSONFieldStringArray(t, batchSendResBody0, "event_ids")
+			nextBatchID0 := client.GetJSONFieldStr(t, batchSendResBody0, "next_batch_id")
+
+			contextRes0 := alice.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "context", historicalEventIDs0[1]}, client.WithContentType("application/json"))
+			contextResBody0 := client.ParseJSON(t, contextRes0)
+			stateJsonResult0 := gjson.GetBytes(contextResBody0, "state")
+			stateFromContext0 := stateJsonResult0.Array()
+
+			logrus.WithFields(logrus.Fields{
+				"stateFromContext": stateFromContext0,
+				"len":              len(stateFromContext0),
+			}).Error("gaweegegegwegawagewg 0")
+
+			if len(stateFromContext0) == 0 {
+				t.Fatalf("Expected some state events in the context response for historical event in first batch but saw %d: %s", len(stateFromContext0), stateFromContext0)
+			}
+
+			// Insert another older batch of backfilled history from the same user.
+			// Make sure the meta data and joins still work on the subsequent batch
+			insertTime1 := timeAfterEventBefore
+			batchSendRes1 := batchSendHistoricalMessages(
+				t,
+				as,
+				roomID,
+				eventIdBefore,
+				nextBatchID0,
+				createJoinStateEventsForBackfillRequest([]string{virtualUserID}, insertTime1),
+				createMessageEventsForBackfillRequest([]string{virtualUserID}, insertTime1, 3),
+				// Status
+				200,
+			)
+			batchSendResBody1 := client.ParseJSON(t, batchSendRes1)
+			historicalEventIDs1 := client.GetJSONFieldStringArray(t, batchSendResBody1, "event_ids")
+
+			contextRes1 := alice.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "context", historicalEventIDs1[1]}, client.WithContentType("application/json"))
+			contextResBody1 := client.ParseJSON(t, contextRes1)
+			stateJsonResult1 := gjson.GetBytes(contextResBody1, "state")
+			stateFromContext1 := stateJsonResult1.Array()
+
+			logrus.WithFields(logrus.Fields{
+				"stateFromContext": stateFromContext1,
+				"len":              len(stateFromContext1),
+			}).Error("gaweegegegwegawagewg 1")
+
+			if len(stateFromContext1) == 0 {
+				t.Fatalf("Expected some state events in the context response for historical event in second batch but saw %d: %s", len(stateFromContext1), stateFromContext1)
```

Comment:
This currently doesn't work because the historical messages don't have proper `state_groups`.

> Element is having trouble pulling in the avatar/displayname because the floating outlier member event state (from `state_events_at_start`) is not resolved into the current room state as expected but also the historical events don't have proper `state_groups`.
> 
> Instead of worrying about `state_groups`, we could just pull from profile as a fallback which is also noted in https://matrix.org/docs/spec/client_server/latest#lazy-loading-room-members ("Clients which do not support tab-completion may instead pull in profiles for arbitrary users (e.g. read receipts, typing notifications) on demand by querying the room state or /profile.")
> 
> *-- https://gitlab.com/gitterHQ/webapp/-/merge_requests/2229#note_683532091*